### PR TITLE
Stacked Borrows: add code for the two core operations

### PR DESCRIPTION
**Pull request by RalfJung**
Namely, accessing memory and dereferencing a pointer.

As requested by @ubsan

### CODE REVIEW

**Comment by gnzlbg** [comment]
You might want to include in a different PR a glossary at the beginning of this file with a brief description of what each technical term introduced here means. I think that would improve the document.

**Comment by RalfJung** [comment]
Could you give a list of terms that you think need explanation? That would help a lot.

**Comment by gnzlbg** [comment]
Sure, I was thinking, something like this

* `Timestamp`: a monotonically increasing counter. 

* `place`:  a memory location (maybe link to the general glossary here).

* `Borrow`: a per-pointer tag `tag` indicating whether the memory referred by the point is uniquely referred by this pointer or aliased by other pointers.

* `CallId`: Per-call frame timestamp.

*  `FnBarrier`: <it is unclear to me what this does, looking at the document, it is unclear to me when these are pushed or why, only when they are popped>

* `Stack`: a per-place `stack` of borrows where the borrow with the largest `Timestamp` is at the top. Each borrow is either unique, was created from a raw pointer, or <barrier> whatever that does. The `Stack` might be frozen since a particular `Timestamp`.

etc.

I think it might be a good idea to give a brief definition of each of the most commonly use terms at the beginning. So that when the reader goes through the text it already has an intuition for what they are. Particularly since some types are used before they are introduced (e.g. `Raw`, `CallId`, etc.).

I had some problems reading the text sequentially, finding myself trying to jump back and forth a couple of times trying to find an explanation for what some types are supposed to represent.

**Comment by RalfJung** [comment]
Thanks!

Notice that this document is not intended as a guide. I expect that people first read the blog posts, and then come here as a reference for the details on one operation.  But I agree adding a brief description (doc-comment style) to each type makes sense.

**Comment by RalfJung** [comment]
I have added doc-comments to the types and enum variants. I hope that helps.

File: wip/stacked-borrows.md [hunk 1]
```diff
@@ -15,31 +19,50 @@ Every pointer value has a *tag* (in addition to the location in memory that the
...
     borrows: Vec<BorStackItem>, // used as a stack; never empty
     frozen_since: Option<Timestamp>, // virtual frozen "item" on top of the stack
 }
 
-// Extra per-call-frame state
+/// Extra per-call-frame state: an ID uniquely identifying a stack frame.
 type CallId = nat;
 
-// Extra global state
+/// Extra global state: the next unused `Timestamp`, as well as the next unused `CallId`.
```
Comment (gnzlbg) [inline wip/stacked-borrows.md]: What does "unused" mean here?

File: wip/stacked-borrows.md [hunk 2]
```diff
@@ -120,14 +143,15 @@ impl Tracking {
...
+In the following, we pretend there exists a function `barrier_is_active(id)` that can check this.
 
 **Note**: Miri uses a slightly more complex system with a `HashSet<CallId>` tracking the set of active `CallId`; that is just an optimization to avoid having to scan the call stack all the time.
 
 ### Allocating memory
 
 When allocating memory, we have to initialize the `Stack` associated with the new locations, and we have to choose a `Borrow` (a tag) for the initial pointer to this memory.
 
-For most memory, the stack of each freshly allocated memory location is `Stack { borrows: vec![Shr], frozen_since: None }`, and the initial pointer to that memory has tag `Shr(None)`.
+For most memory, the stack of each freshly allocated memory location is `Stack { borrows: vec![Raw], frozen_since: None }`, and the initial pointer to that memory has tag `Alias(None)`.
```
Comment (gnzlbg) [inline wip/stacked-borrows.md]: Do we also carry the alignment of that memory ? The section below on dereferencing raw pointers states that `size_of_val` is checked, but it says nothing about checking that the pointer is at an alignment boundary of `T`.

Comment (gnzlbg) [reply wip/stacked-borrows.md]: Or is that implied as part of "the pointer must be dereferenceable" ?

File: wip/stacked-borrows.md [hunk 3]
```diff
@@ -176,27 +325,52 @@ we determine the extent of memory that this place covers using `size_of_val` and
    Remember the position of the item matching the tag in the stack.
```
Comment (gnzlbg) [inline wip/stacked-borrows.md]: typo: "we determine the extent of memory" should start with a capital letter: "We determine ..."

File: wip/stacked-borrows.md (hunk 1 above)
Comment (RalfJung) [reply wip/stacked-borrows.md]: "not used"? I don't get the question.

File: wip/stacked-borrows.md (hunk 2 above)
Comment (RalfJung) [reply wip/stacked-borrows.md]: Stacked Borrows does not care about alignment at all, and that's why it is not mentioned.

File: wip/stacked-borrows.md (hunk 3 above)
Comment (RalfJung) [reply wip/stacked-borrows.md]: Thanks! Fixed.

File: wip/stacked-borrows.md (hunk 2 above)
Comment (gnzlbg) [reply wip/stacked-borrows.md]: Miri complains both on the creation of an unaligned reference and on the pointer dereference below, so I wonder how / where this is tracked ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4db5abaec7e73cacba66f65c683e9cac)):

```
use core::{arch::x86_64::__m128, mem::transmute};
#[derive(Copy, Clone)]
#[repr(align(8))]
struct U8(u8);
fn main() {
    unsafe {
        type U = U8;
        let a: [U8; 16] = [U8(0); 16];
        let b: *const __m128 = &a as &[U] as *const [U] as *const U as *const __m128;
        let _c: &__m128 = transmute(b);
        let _d = *b;
    }
}
```

Comment (RalfJung) [reply wip/stacked-borrows.md]: The former is the validity invariant, and the latter is the normal rule for a memory access.

This question would be answered by a proper definition of MIR semantics (and, in particular, places), which I will write... eventually.^^  But it is not part of Stacked Borrows and IMO should not be answered here.

Comment (gnzlbg) [reply wip/stacked-borrows.md]: > But it is not part of Stacked Borrows and IMO should not be answered here.

Fair enough, and thanks!

**Pull request merged**